File: apps/mobile/modules/t3-lan-discovery/android/src/main/java/expo/modules/t3landiscovery/T3LanDiscoveryModule.kt
package expo.modules.t3landiscovery

import android.content.Context
import android.net.nsd.NsdManager
import android.net.nsd.NsdServiceInfo
import android.net.wifi.WifiManager
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.net.InetAddress
import java.util.concurrent.atomic.AtomicBoolean

class T3LanDiscoveryModule : Module() {
  private var nsdManager: NsdManager? = null
  private var discoveryListener: NsdManager.DiscoveryListener? = null
  private var multicastLock: WifiManager.MulticastLock? = null
  private val discovering = AtomicBoolean(false)
  private val resolvingServices = mutableSetOf<String>()

  override fun definition() = ModuleDefinition {
    Name("T3LanDiscovery")

    Events("onLanDiscoveryEvent")

    AsyncFunction("start") { serviceType: String ->
      val context = appContext.reactContext ?: throw Exceptions.ReactContextLost()
      startDiscovery(context, normalizeServiceType(serviceType))
    }

    AsyncFunction("stop") {
      stopDiscovery()
    }
  }

  private fun normalizeServiceType(serviceType: String): String {
    val trimmed = serviceType.trim().ifEmpty { "_t3code._tcp" }
    val base = trimmed.removePrefix("_").removeSuffix("._tcp").removeSuffix(".")
    return "_${base}._tcp."
  }

  private fun startDiscovery(context: Context, serviceType: String) {
    stopDiscovery()

    val wifi = context.applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
    multicastLock =
      wifi.createMulticastLock("t3-lan-discovery").apply {
        setReferenceCounted(true)
        acquire()
      }

    nsdManager = context.getSystemService(Context.NSD_SERVICE) as NsdManager
    discovering.set(true)
    resolvingServices.clear()

    discoveryListener =
      object : NsdManager.DiscoveryListener {
        override fun onDiscoveryStarted(regType: String) {}

        override fun onServiceFound(serviceInfo: NsdServiceInfo) {
          if (!discovering.get()) {
            return
          }
          val serviceName = serviceInfo.serviceName ?: return
          if (resolvingServices.contains(serviceName)) {
            return
          }
          resolvingServices.add(serviceName)

          nsdManager?.resolveService(
            serviceInfo,
            object : NsdManager.ResolveListener {
              override fun onResolveFailed(info: NsdServiceInfo, errorCode: Int) {
                resolvingServices.remove(serviceName)
              }

              override fun onServiceResolved(info: NsdServiceInfo) {
                resolvingServices.remove(serviceName)
                if (!discovering.get()) {
                  return
                }
                val host = info.host?.hostAddress ?: return
                val port = info.port
                if (port <= 0) {
                  return
                }
                sendEvent(
                  "onLanDiscoveryEvent",
                  mapOf(
                    "type" to "resolved",
                    "serviceName" to (info.serviceName ?: serviceName),
                    "host" to host,
                    "port" to port,
                  ),
                )
              }
            },
          )
        }

        override fun onServiceLost(serviceInfo: NsdServiceInfo) {
          val serviceName = serviceInfo.serviceName ?: return
          resolvingServices.remove(serviceName)
          sendEvent(
            "onLanDiscoveryEvent",
            mapOf(
              "type" to "lost",
              "serviceName" to serviceName,
            ),
          )
        }

        override fun onDiscoveryStopped(serviceType: String) {}

        override fun onStartDiscoveryFailed(serviceType: String, errorCode: Int) {
          sendEvent(
            "onLanDiscoveryEvent",
            mapOf(
              "type" to "error",
              "message" to "LAN discovery failed to start (code $errorCode).",
            ),
          )
          stopDiscovery()
        }

        override fun onStopDiscoveryFailed(serviceType: String, errorCode: Int) {
          sendEvent(
            "onLanDiscoveryEvent",
            mapOf(
              "type" to "error",
              "message" to "LAN discovery failed to stop (code $errorCode).",
            ),
          )
        }
      }

    nsdManager?.discoverServices(serviceType, NsdManager.PROTOCOL_DNS_SD, discoveryListener)
  }

  private fun stopDiscovery() {
    discovering.set(false)
    resolvingServices.clear()

    val listener = discoveryListener
    if (listener != null) {
      try {
        nsdManager?.stopServiceDiscovery(listener)
      } catch (_: IllegalArgumentException) {
      }
    }
    discoveryListener = null
    nsdManager = null

    multicastLock?.let { lock ->
      if (lock.isHeld) {
        lock.release()
      }
    }
    multicastLock = null
  }
}

File: apps/mobile/modules/t3-lan-discovery/ios/T3LanDiscoveryModule.swift
import ExpoModulesCore
import Foundation

private let defaultServiceType = "_t3code._tcp."

private func normalizeServiceType(_ serviceType: String) -> String {
  let trimmed = serviceType.trimmingCharacters(in: .whitespacesAndNewlines)
  if trimmed.isEmpty {
    return defaultServiceType
  }
  var base = trimmed
  if base.hasPrefix("_") {
    base.removeFirst()
  }
  if base.hasSuffix("._tcp.") {
    return "_\(base)"
  }
  if base.hasSuffix("._tcp") {
    return "_\(base)."
  }
  return "_\(base)._tcp."
}

public class T3LanDiscoveryModule: Module, NetServiceBrowserDelegate, NetServiceDelegate {
  private var browser: NetServiceBrowser?
  private var resolvingServices: [String: NetService] = [:]

  public func definition() -> ModuleDefinition {
    Name("T3LanDiscovery")

    Events("onLanDiscoveryEvent")

    AsyncFunction("start") { (serviceType: String) in
      self.startBrowsing(serviceType: serviceType)
    }

    AsyncFunction("stop") {
      self.stopBrowsing()
    }
  }

  private func startBrowsing(serviceType: String) {
    stopBrowsing()
    let browser = NetServiceBrowser()
    browser.delegate = self
    browser.searchForServices(ofType: normalizeServiceType(serviceType), inDomain: "local.")
    self.browser = browser
  }

  private func stopBrowsing() {
    browser?.stop()
    browser?.delegate = nil
    browser = nil
    resolvingServices.removeAll()
  }

  public func netServiceBrowser(
    _ browser: NetServiceBrowser,
    didFind service: NetService,
    moreComing: Bool
  ) {
    guard let serviceName = service.name as String? else {
      return
    }
    if resolvingServices[serviceName] != nil {
      return
    }
    service.delegate = self
    resolvingServices[serviceName] = service
    service.resolve(withTimeout: 5)
  }

  public func netServiceBrowser(
    _ browser: NetServiceBrowser,
    didRemove service: NetService,
    moreComing: Bool
  ) {
    guard let serviceName = service.name as String? else {
      return
    }
    resolvingServices.removeValue(forKey: serviceName)
    sendEvent(
      "onLanDiscoveryEvent",
      [
        "type": "lost",
        "serviceName": serviceName,
      ]
    )
  }

  public func netServiceBrowser(_ browser: NetServiceBrowser, didNotSearch errorDict: [String: NSNumber]) {
    sendEvent(
      "onLanDiscoveryEvent",
      [
        "type": "error",
        "message": "LAN discovery failed to start.",
      ]
    )
    stopBrowsing()
  }

  public func netServiceDidResolveAddress(_ sender: NetService) {
    guard let serviceName = sender.name as String? else {
      return
    }
    guard sender.port > 0 else {
      return
    }
    guard let host = sender.hostName?.trimmingCharacters(in: CharacterSet(charactersIn: ".")) else {
      return
    }

    sendEvent(
      "onLanDiscoveryEvent",
      [
        "type": "resolved",
        "serviceName": serviceName,
        "host": host,
        "port": sender.port,
      ]
    )
  }

  public func netService(_ sender: NetService, didNotResolve errorDict: [String: NSNumber]) {
    guard let serviceName = sender.name as String? else {
      return
    }
    resolvingServices.removeValue(forKey: serviceName)
  }
}